### ISSUE TITLE
Consider removing Marshal.ReleaseComObject(...) calls

### ISSUE THREAD

**Issue by govert** [christso/DAX-Drill#1]
I take a look at Excel add-ins using Excel-DNA from time to time, and found your library. Open source libraries like this serve as great examples and inspiration for other too, so thank you for posting your work online.

There is a lot of confusion and misinformation about making calls to `Marshal.ReleaseComObject()` when dealing with COM libraries from managed code. I'm now quite sure that calling `Marshal.ReleaseComObject()` is never needed, and can often do harm.

My recent answers in this discussion explain the issue as I understand it, and includes some links: https://social.msdn.microsoft.com/Forums/office/en-US/e6faed91-32ff-46c4-b1d5-c9575c40b4a2/cannot-release-excelrange-object?forum=exceldev

You might want to have a look, and check whether your library can be simplified a bit by removing these calls.

**Comment by christso** [comment]
Thanks for the feedback! The `Marshal.ReleaseComObject()` pattern was really slowing me down as I needed to ensure that I did not prematurely release COM objects, i.e. if I release a COM object then I needed to ensure that the object won't be used later on. That will make things a lot easier if `Marshal.ReleaseComObject()` is no longer needed.

I originally followed the advice in the Add-In Express blog:
https://www.add-in-express.com/creating-addins-blog/2013/11/05/release-excel-com-objects/

They also advise I don't use `foreach` loops with COM objects, so my code uses the less readable `for` loop. That blog post is still very active.

I'll clean up my code in the next few days.
